heudiconv/parser.py

```diff
@@ -60,58 +62,87 @@ def find_files(regex, topdir=op.curdir, exclude=None,
+    list.
 
-    # cannot use TempDirs since will trigger cleanup with __del__
-    tmpdir = tempdirs('heudiconvDCM')
+    When contents of fl appear to be an unpackable archive, the contents are
+    extracted into utils.TempDirs(f'heudiconvDCM') and the mode of all 
+    extracted files is set to 700.
 
-    sessions = defaultdict(list)
-    session = 0
-    if not isinstance(fl, (list, tuple)):
+    When contents of fl are a list of files, they are treated as a single
+    session.
+    """
+    input_list_of_unpacked = any(fl) and all(
+        not t.endswith(_UNPACK_FORMATS) for t in fl
+        )
+    
+    if not (
+        input_list_of_unpacked or
+        all(t.endswith(_UNPACK_FORMATS) for t in fl)
+        ):
+        raise ValueError("Some but not all input files are archives.")
+    
+    sessions: typing.DefaultDict[
+        typing.Optional[int], 
+        typing.List[str]
+        ] = defaultdict(list)
+    
+    if not isinstance(fl, list):
         fl = list(fl)
 
-    # needs sorting to keep the generated "session" label deterministic
-    for i, t in enumerate(sorted(fl)):
-        # "classical" heudiconv has that heuristic to handle multiple
-        # tarballs as providing different sessions per each tarball
-        if not tarfile.is_tarfile(t):
-            sessions[None].append(t)
-            continue
-
-        tf = tarfile.open(t)
-        # check content and sanitize permission bits
-        tmembers = tf.getmembers()
-        for tm in tmembers:
-            tm.mode = 0o700
-        # get all files, assemble full path in tmp dir
-        tf_content = sorted([m.name for m in tmembers if m.isfile()])
-        # store full paths to each file, so we don't need to drag along
-        # tmpdir as some basedir
-        sessions[session] = [op.join(tmpdir, f) for f in tf_content]
-        session += 1
-        # extract into tmp dir
-        tf.extractall(path=tmpdir, members=tmembers)
-
-    if session == 1:
-        # we had only 1 session, so no really multiple sessions according
-        # to classical 'heudiconv' assumptions, thus just move them all into
-        # None
-        sessions[None] += sessions.pop(0)
+    if input_list_of_unpacked:
+        sessions[None] = fl
+    else:
+        # keep track of session manually to ensure that the variable is bound
+        # when it is used after the loop (e.g., consider situation with
+        # fl being empty)
+        session = 0
+        # needs sorting to keep the generated "session" label deterministic
+        for _, t in enumerate(sorted(fl)):
+            # Each file extracted must be associated with the proper session,
+            # but the high-level shutil does not have a way to list the files
+            # contained within each archive. So, files are temporarily
+            # extracted into unique tempdirs
+            # cannot use TempDirs since will trigger cleanup with __del__
+            tmpdir = tempdirs(prefix="heudiconvDCM")
```

Comment:
make that tmpdir `0o700` right away here so no files extracted would have any moment in time when they are accessible to non-authorized user (i.e. until chmoding below)?

Reply:
Good point! This seems especially relevant considering that the files are unpacked into a generic tmp directory, which could be a surprising risk to privacy.